### PR TITLE
IIS: Identify if a request has a body 

### LINKED ISSUE #24175: Identifying if a request has a body
It's difficult to check if an HTTP request has a body without actually calling read on that body. The caller might not want to call read because that can have side-effects like 100-continue, throwing, etc.. YARP is one example where we need to know if there is a request body before calling read (https://github.com/microsoft/reverse-proxy/pull/336) because YARP doesn't call read itself, it generates an HttpContent object for HttpClient to call read on.

In HTTP/1.0 a request body must be indicated by a Content-Length.
In HTTP/1.1 a request body must be indicated by a Content-Length or Transfer-Encoding: chunked.
In HTTP/2 (and 3?) Content-Length is optional and Transfer-Encoding: chunked is prohibited. You only know there is no body if you've received the END_STREAM flag.

Http.Sys/IIS have a history of faking HTTP/2 as HTTP/1.1. For requests with bodies and no Content-Length it will add a fake Transfer-Encoding: chunked header.

For YARP it's easy to check for the Content-Length and chunked headers, but checking if an HTTP/2 request has a body on Kestrel isn't possible today.

Proposal: Kestrel should also add a fake header. Either:
- Content-Length: 0 for any request without a body, or
- Transfer-Encoding: chunked for HTTP/2+ request with a body, same as Http.Sys/IIS.

### PR DESCRIPTION
**Pull request by Tratcher**
Fixes #24175. This is the IIS portion of the feature. IIS is easy because Http.Sys adds a fake Transfer-Encoding: chunked header even for HTTP/2 requests.

### CODE REVIEW

File: src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/Http2Tests.cs [hunk 1]
```diff
@@ -0,0 +1,1017 @@
...
+
+                    dataFrame = await h2Connection.ReceiveFrameAsync();
+                    Http2Utilities.VerifyDataFrame(dataFrame, 1, endOfStream: true, length: 0);
+
+                    h2Connection.Logger.LogInformation("Connection stopped.");
+                })
+                .Build().RunAsync();
+        }
+        /*
+        [ConditionalFact(Skip = "https://github.com/dotnet/aspnetcore/issues/17420")]
```
Comment (jkotalik) [inline src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/Http2Tests.cs:293]: ?

File: src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/Http2Tests.cs [hunk 2]
```diff
@@ -0,0 +1,1017 @@
...
+using Microsoft.AspNetCore.Testing;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Microsoft.Net.Http.Headers;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Server.IIS.FunctionalTests.InProcess
+{
+    [Collection(PublishedSitesCollection.Name)]
+    public class Http2Tests : IISFunctionalTestBase
```
Comment (jkotalik) [inline src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/Http2Tests.cs:24]: I'm worried about test time with all of these tests.

**Review by jkotalik — APPROVED** [review]

File: src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/Http2Tests.cs (hunk 2 above)
Comment (Tratcher) [reply src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/Http2Tests.cs:24]: These tests average 2.5 seconds each, for a total of 43 seconds.

These tests don't need anything special except https, can we create an https fixture they can share?

File: src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/Http2Tests.cs (hunk 1 above)
Comment (Tratcher) [reply src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/Http2Tests.cs:293]: Copied from HttpSys, but you have reset and goaway tests elsewhere. Removed.

File: src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/Http2Tests.cs (hunk 2 above)
Comment (jkotalik) [reply src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/Http2Tests.cs:24]: You can, but no need to block on doing that to implement it. Adding fixtures here is kind of tricky.

Comment (Tratcher) [reply src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/Http2Tests.cs:24]: https://github.com/dotnet/aspnetcore/issues/25384

File: src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/Http2Tests.cs [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
```
Comment (halter73) [inline src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/Http2Tests.cs:24]: Are there any HTTP/1.x tests for RequestCanHaveBody?

File: src/Servers/IIS/IIS/src/Core/IISHttpContext.cs [hunk 4]
```diff
@@ -250,6 +253,20 @@ public string ReasonPhrase
 
         internal IISHttpServer Server => _server;
 
+        private bool CheckRequestCanHaveBody()
+        {
+            // Http/1.x requests with bodies require either a Content-Length or Transfer-Encoding header.
+            // Note Http.Sys adds the Transfer-Encoding: chunked header to HTTP/2 requests with bodies for back compat.
+            // Transfer-Encoding takes priority over Content-Length.
+            string transferEncoding = RequestHeaders[HttpKnownHeaderNames.TransferEncoding];
+            if (string.Equals("chunked", transferEncoding?.Trim(), StringComparison.OrdinalIgnoreCase))
```
Comment (halter73) [inline src/Servers/IIS/IIS/src/Core/IISHttpContext.cs:262]: Does Http.sys not trim header values?

Comment (Tratcher) [reply src/Servers/IIS/IIS/src/Core/IISHttpContext.cs:262]: It probably does, this code is just defensive.

**Comment by Tratcher** [comment]
I'm struggling with the IIS test failures. It only fails in IIS, not express, and only on the CI. It seems to get back a 200 response with no body when we're expecting a body to be echo'd.

I'm going to try updating the CI agents. I know we're using obsolete versions of Win10. https://github.com/dotnet/aspnetcore/pull/25627

File: src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/Http2Tests.cs (hunk 3 above)
Comment (Tratcher) [reply src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/Http2Tests.cs:24]: I've added a few

**Comment by Tratcher** [comment]
@Pilchie please merge for rc2.

**Comment by Pilchie** [comment]
Approved for .NET 5 RC2.

**Pull request merged**